witness/golang/omniwitness/usbarmory/internal/storage/persistance.go

```diff
@@ -0,0 +1,293 @@
+	p.mapWriteToken = t
+
+	// Precalculate the list of available slots.
+	slotState := make([]bool, p.part.NumSlots())
+	for _, idx := range p.idToSlot {
+		if idx == mappingConfigSlot {
+			return errors.New("internal-error, reserved slot 0 has been used")
+		}
+		slotState[idx] = true
+	}
+
+	// Slot 0 is reserved for the mapping config, so mark it used here:
+	slotState[mappingConfigSlot] = true
+
+	p.freeSlots = make([]uint, 0, p.part.NumSlots())
+	for idx, used := range slotState {
+		if !used {
+			p.freeSlots = append(p.freeSlots, uint(idx))
+		}
+	}
+	return nil
+}
+
+// storeMap writes the current logID -> slot map to storage.
+// Must be called with p.mu locked.
+func (p *SlotPersistence) storeMap() error {
+	smRaw, err := yaml.Marshal(p.idToSlot)
+	if err != nil {
+		return fmt.Errorf("failed to marshal mapping: %v", err)
+	}
+	if err := p.mapSlot.CheckAndWrite(p.mapWriteToken, smRaw); err != nil {
+		return fmt.Errorf("failed to store mapping: %v", err)
+	}
+	// TODO(al): CheckAndWrite should return the next token rather than us knowing
+	// how the token changes after a successful write.
+	p.mapWriteToken++
+	return nil
+}
+
+// addLog assigns a slot to a new log ID.
+// Must be called with p.mu locked.
+func (p *SlotPersistence) addLog(id string) (uint, error) {
+	if idx, ok := p.idToSlot[id]; ok {
+		return idx, nil
+	}
+	if len(p.freeSlots) == 0 {
+		return 0, errors.New("no free slot available")
+	}
+	f := p.freeSlots[0]
+	p.freeSlots = p.freeSlots[1:]
+	p.idToSlot[id] = f
+	return f, p.storeMap()
+}
+
+// Init sets up the persistence layer. This should be idempotent,
+// and will be called once per process startup.
+func (p *SlotPersistence) Init() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	s, err := p.part.Open(mappingConfigSlot)
+	if err != nil {
+		return fmt.Errorf("failed to open mapping slot: %v", err)
+	}
+	p.mapSlot = s
+	if err := p.populateMap(); err != nil {
+		return fmt.Errorf("failed to populate logID → slot map: %v", err)
+	}
+	return nil
+}
+
+// Logs returns the IDs of all logs that have checkpoints that can
+// be read.
+func (p *SlotPersistence) Logs() ([]string, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	r := make([]string, 0, len(p.idToSlot))
+	for k := range p.idToSlot {
+		r = append(r, k)
+	}
+	return r, nil
+}
+
+// ReadOps returns read-only operations for the given log ID. This
+// method only makes sense for IDs returned by Logs().
+func (p *SlotPersistence) ReadOps(logID string) (persistence.LogStateReadOps, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	i, ok := p.idToSlot[logID]
+	if !ok {
+		// TODO(al): work around undocumented assumptions in the storage interface;
+		// we have to return a ReadOp even if we don't know about the specified logID.
+		// the ReadOps we return must then return an error of codes.NotFound when GetLatest
+		// is called so that the witness will eventually call WriteOps/Update with the
+		// same logID and create the record.
+		return &slotOps{}, nil
```

Comment:
Could do that, but it might be worth thinking about whether this idea of returning an ops we already know is going to return errors is the pattern we want to double down on.
Leaving the TODO and extra 3 lines to work around seems reasonable until we've figured it out?

Reply:
Sure, let's work out what we want the persistence contract to be.